sdk/src/types/block/semantic.rs

```diff
@@ -327,6 +327,82 @@ pub fn semantic_validation(
             Output::Delegation(output) => (output.amount(), None, None),
         };
 
+        if let Some(unlock_conditions) = created_output.unlock_conditions() {
+            // Check the possibly restricted address-containing conditions
+            let addresses = unlock_conditions
+                .iter()
+                .filter_map(|uc| match uc {
+                    UnlockCondition::Address(uc) => Some(uc.address()),
+                    UnlockCondition::Expiration(uc) => Some(uc.return_address()),
+                    UnlockCondition::StateControllerAddress(uc) => Some(uc.address()),
+                    UnlockCondition::GovernorAddress(uc) => Some(uc.address()),
+                    _ => None,
+                })
+                .filter_map(Address::as_restricted_opt);
+            for address in addresses {
+                if created_output.native_tokens().map(|t| t.len()).unwrap_or_default() > 0
+                    && !address
+                        .allowed_capabilities()
+                        .has_capability(AddressCapabilityFlag::OutputsWithNativeTokens)
```

Comment:
I feel like we should have a `RestrictedAddress::has_capability` directly

Reply:
added